coderd/notifications/notifications_test.go

```diff
@@ -938,53 +985,263 @@ func TestNotificationTemplates_Golden(t *testing.T) {
+						return false
+					}
+
+					if err = cl.Close(); err != nil {
+						t.Logf("smtp didn't close properly: %s", err)
+						return false
+					}
 
-			require.Equal(t, string(wantBody), body, fmt.Sprintf("rendered template body does not match golden file. If this is expected, %s", hint))
-			require.Equal(t, string(wantTitle), title, fmt.Sprintf("rendered template title does not match golden file. If this is expected, %s", hint))
+					return true
+				}, testutil.WaitShort, testutil.IntervalFast)
+
+				smtpCfg := defaultNotificationsConfig(database.NotificationMethodSmtp)
+				smtpCfg.SMTP = smtpConfig
+
+				smtpManager, err := notifications.NewManager(
+					smtpCfg,
+					*db,
+					defaultHelpers(),
+					createMetrics(),
+					logger.Named("manager"),
+				)
+				require.NoError(t, err)
+
+				smtpManager.Run(ctx)
+				time.Sleep(time.Second)
+
+				notificationCfg := defaultNotificationsConfig(database.NotificationMethodSmtp)
+
+				smtpEnqueuer, err := notifications.NewStoreEnqueuer(
+					notificationCfg,
+					*db,
+					defaultHelpers(),
+					logger.Named("enqueuer"),
+					quartz.NewReal(),
+				)
+				require.NoError(t, err)
+
+				_, err = smtpEnqueuer.EnqueueWithData(
+					ctx,
+					user.ID,
+					tc.id,
+					tc.payload.Labels,
+					tc.payload.Data,
+					user.Username,
+					user.ID,
+				)
+				require.NoError(t, err)
+
+				// Wait for the message to be fetched
+				var msg *mocksmtp.Message
+				require.Eventually(t, func() bool {
+					msg = backend.LastMessage()
+					return msg != nil
+				}, testutil.WaitShort, testutil.IntervalFast)
+				require.NotNil(t, msg, "want a message to be sent")
+
+				body := normalizeGoldenEmail([]byte(msg.Contents))
+
+				err = smtpManager.Stop(ctx)
+				require.NoError(t, err)
+
+				partialName := strings.Split(t.Name(), "/")[1]
+				goldenFile := filepath.Join("testdata", "rendered-templates", "smtp", partialName+".html.golden")
+				if *updateGoldenFiles {
+					err = os.MkdirAll(filepath.Dir(goldenFile), 0o755)
+					require.NoError(t, err, "want no error creating golden file directory")
+					err = os.WriteFile(goldenFile, body, 0o600)
+					require.NoError(t, err, "want no error writing body golden file")
+					return
+				}
+
+				wantBody, err := os.ReadFile(goldenFile)
+				require.NoError(t, err, fmt.Sprintf("missing golden notification body file. %s", hint))
+				require.Equal(t, string(wantBody), string(body), fmt.Sprintf("smtp notification does not match golden file. If this is expected, %s", hint))
+			})
+
+			t.Run("webhook", func(t *testing.T) {
+				t.Parallel()
+				// nolint:gocritic // Unit test.
+				ctx := dbauthz.AsSystemRestricted(testutil.Context(t, testutil.WaitSuperLong))
+
+				// Spin up the mock webhook server
+				server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					w.WriteHeader(http.StatusOK)
+
+					body, err := io.ReadAll(r.Body)
+					require.NoError(t, err)
```

Comment:
Nit: Since the handler runs in a separate goroutine, we can't use `require` here as it calls `t.FailNow` (https://pkg.go.dev/testing#F.FailNow).

You could either move handling back to test routine via channel or use assert+return instead. We should also make sure that the handler is done before we exit the test, especially if we keep logic in it (i.e. verify if Close waits for all handlers to complete or immediately stops the server and ignores the handlers continued execution).

Reply:
Fixed